libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,336 @@
+                .pollDelay(POLL_DELAY_MILLIS, TimeUnit.MILLISECONDS)
+                .until(() -> messages1.size() + messages2.size() == 4);
+
+            assertEquals(2, messages1.size());
+            assertEquals(2, messages2.size());
+
+            dispatcher.unsubscribe(subscription1);
+            dispatcher.unsubscribe(subscription2);
+        }
+    }
 
+    @Test
+    public void whenSubscribingAsynchronouslyWithQueueGroups_thenOnlyOneMessageHandlerInTheGroupShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            List<Message> messages = new ArrayList<>();
+
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            Subscription qSub1 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+            Subscription qSub2 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+
+            natsConnection.publish("mySubject", "data".getBytes());
+
+            // Simulate real world time for messages to propagate.
+            Awaitility.await()
+                .atMost(WAIT_AT_MOST_MILLIS, TimeUnit.MILLISECONDS)
+                .pollDelay(POLL_DELAY_MILLIS, TimeUnit.MILLISECONDS)
+                .until(() -> messages.size() == 1);
+
+            assertEquals(1, messages.size());
+            dispatcher.unsubscribe(qSub1);
+            dispatcher.unsubscribe(qSub2);
+        }
     }
 
+    @Test
+    public void whenMessagesAreExchangedViaPublish_thenResponsesMustBeReceivedWithSecondarySubscription() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            Subscription replySideSubscription1 = natsConnection.subscribe("publishSubject");
+            Subscription replySideSubscription2 = natsConnection.subscribe("publishSubject");
+
+            Subscription publishSideSubscription = natsConnection.subscribe("replyToSubject");
+            natsConnection.publish("publishSubject", "replyToSubject", "Please respond!".getBytes());
+
+            Message message = replySideSubscription1.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            assertEquals("Please respond!", new String(message.getData()));
+            natsConnection.publish(message.getReplyTo(), "Message Received By Subscription 1".getBytes());
+
+            message = replySideSubscription2.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            assertEquals("Please respond!", new String(message.getData()));
+            natsConnection.publish(message.getReplyTo(), "Message Received By Subscription 2".getBytes());
+
+            int responsesFrom1 = 0;
+            int responsesFrom2 = 0;
+            message = publishSideSubscription.nextMessage(TIMEOUT_MILLIS);
+            assertNotNull(message, "No message!");
+            String response = new String(message.getData());
+            assertTrue(response.contains("Message Received"));
```

Comment:
For complex assertions we use AssertJ, which has helpers like:
> assertThat(response).contains("..");

And usually if we use AssertJ at all we would use it everywhere for consistency

Reply:
I give up. Is this really that big of a deal, it would require going back through every single test? These assertions are not complex.

Reply:
I can link you to our guidelines on this and show you what my reference point is. We have guidelines in order to achieve consistency between all of the articles on the site.

Reply:
Alternatively we can wait until final review (when another person checks over the PR and article) and let them decide any points you would like not to do